File: packages/sdk-typescript/src/event-builder.ts
import { generateId } from '@act/core';

export interface BuildEventParams {
  eventType: string;
  actor: { actorId: string; keyId: string };
  tenant: string | { not_applicable: true; reason: string };
  subject: {
    kind: 'artifact' | 'transformation' | 'attestation' | 'policy' | 'evidence';
    artifact_id?: string;
    version_id?: string;
    artifact_type?: string;
  };
  causalParents?: { event_id: string; relation?: string }[];
  contentDescriptors?: Record<string, unknown>[];
  policyContext?:
    { policy_id: string; policy_version: string } | { not_applicable: true; reason: string };
  payload: Record<string, unknown>;
  extensions?: Record<string, unknown>;
  occurredAt?: string;
}

/** Builds an unsigned ACT event payload with protocol defaults filled in. Pass to @act/crypto's signEnvelope to produce a submittable envelope. */
export function buildUnsignedEvent(params: BuildEventParams): Record<string, unknown> {
  return {
    protocol_version: 'act/1.0',
    event_type: params.eventType,
    occurred_at: params.occurredAt ?? new Date().toISOString(),
    actor: { actor_id: params.actor.actorId, key_id: params.actor.keyId },
    tenant: params.tenant,
    subject: params.subject,
    causal_parents: params.causalParents ?? [],
    content_descriptors: params.contentDescriptors ?? [],
    policy_context: params.policyContext ?? {
      not_applicable: true,
      reason: 'no policy configured',
    },
    payload: params.payload,
    extensions: params.extensions ?? {},
  };
}

/** Generates a fresh logical artifact id (UUIDv7). */
export function newArtifactId(): string {
  return generateId();
}

File: sdks/python/src/act_sdk/event_builder.py
"""Ports `packages/sdk-typescript/src/event-builder.ts`."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

from .core.ids import generate_id


def build_unsigned_event(
    event_type: str,
    actor_id: str,
    actor_key_id: str,
    tenant: str | dict[str, Any],
    subject: dict[str, Any],
    payload: dict[str, Any],
    causal_parents: list[dict[str, Any]] | None = None,
    content_descriptors: list[dict[str, Any]] | None = None,
    policy_context: dict[str, Any] | None = None,
    extensions: dict[str, Any] | None = None,
    occurred_at: str | None = None,
) -> dict[str, Any]:
    """Builds an unsigned ACT event payload with protocol defaults filled
    in. Pass to `act_sdk.crypto.sign_envelope` to produce a submittable
    envelope."""
    return {
        "protocol_version": "act/1.0",
        "event_type": event_type,
        "occurred_at": occurred_at or datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
        "actor": {"actor_id": actor_id, "key_id": actor_key_id},
        "tenant": tenant,
        "subject": subject,
        "causal_parents": causal_parents or [],
        "content_descriptors": content_descriptors or [],
        "policy_context": policy_context
        or {"not_applicable": True, "reason": "no policy configured"},
        "payload": payload,
        "extensions": extensions or {},
    }


def new_artifact_id() -> str:
    """Generates a fresh logical artifact id (UUIDv7)."""
    return generate_id()
